Correctly document `rand_distr` features and forward "std" feature to `num-traits`

vks opened the pull request:
Previously, the features were undocumented and it was incorrectly claimed that `no_std` is not supported.

josephlr commented on rand_distr/README.md:
```diff
@@ -19,7 +19,12 @@ It is worth mentioning the [statrs] crate which provides similar functionality
 
-Unlike most Rand crates, `rand_distr` does not currently support `no_std`.
+If the `std` default feature is enabled, `rand_distr` uses the floating point
+functions from `std`. Otherwise, the floating point functions from `num_traits`
+and `libm` are used to support `no_std` environments.
```
Isn't this not true right now? We don't seem to forward the `std` trait to num-traits, so would we even end up using the `std` implementations?

josephlr replied:
ya we need to add `"rand-traits/std"` to this line: https://github.com/rust-random/rand/blob/3a03c9eb5350e03a3f540dba2ee34e0984f2c2c2/rand_distr/Cargo.toml#L24

dhardy replied:
You mean "num-traits/std"?

This actually brings up another question though: *should* we prefer the `std` impls? This might affect portability of results.

josephlr replied:
> You mean "num-traits/std"

Whoops, ya

> This actually brings up another question though: _should_ we prefer the `std` impls?

I feel like we should, if only because that's what num-traits does. Also, preffering `std` over libm is what I would expect. But I don't have that strong a feeling either way.

dhardy replied:
> I feel like we should

Maybe; I don't have a strong opinion on it, though we should probably mention it [in the book](https://rust-random.github.io/book/portability.html) and maybe also try to gather data on where results may differ.

vks replied:
Fair enough, I decided to forward our "std" feature as "num-traits/std", since this is the behavior I would expect.

Of course, this might technically be a value-breaking change. If we really want to have value stability for floating-point distributions, we might want to always use `libm`, because I am not sure that all implementations of the C standard math libraries produce the same results.

dhardy reviewed (commented):
Another point: enabling `std` in `rand` does not enable `std` in `rand_distr`. Perhaps we don't want this though; in any case, we can't without https://github.com/rust-lang/cargo/issues/8832.

vks commented:
> Another point: enabling std in rand does not enable std in rand_distr. Perhaps we don't want this though; in any case, we can't without rust-lang/cargo#8832.

`rand_distr` depends on `rand`, not the other way around. So I think this is as it should be?

vks commented:
> Another point: enabling std in rand does not enable std in rand_distr. Perhaps we don't want this though; in any case, we can't without rust-lang/cargo#8832.

`rand_distr` depends on `rand`, not the other way around. So I think this is as it should be?

vks commented:
Are the changes here ok or should I forgo forwarding "std" for now?

dhardy commented:
> Fair enough, I decided to forward our "std" feature as "num-traits/std", since this is the behavior I would expect.
>
> Of course, this might technically be a value-breaking change. If we really want to have value stability for floating-point distributions, we might want to always use libm, because I am not sure that all implementations of the C standard math libraries produce the same results.

Yes, I'd consider this a breaking change and not for a patch version. Also, regarding the latter point, better value portability sounds desirable. Is there any real cost to using `libm` — e.g. performance or incompatibility? Because if not, is it better to use it?

For now maybe we should add a different flag like `std_math`, and if the next release will be a patch release it should not be on by default.

vks commented:
Ok, I reverted the change to use `std` float functions to keep the status quo and updated the README accordingly.

>  Is there any real cost to using libm — e.g. performance or incompatibility?

There are reports on their issue tracker about `libm` being [twice as slow on some platforms](https://github.com/rust-lang/libm/issues/246). More worryingly, there are reports about [incorrect results due to bugs in the `floor` implementation on `x87` platforms](https://github.com/rust-lang/libm/issues/243). There are also some [accuracy bugs](https://github.com/rust-lang/libm/issues/236).

> For now maybe we should add a different flag like std_math, and if the next release will be a patch release it should not be on by default.

I'm not sure the differences in results would be big enough to justify a feature flag, but let's keep that for another PR.

dhardy commented:
I still don't see a reason not to add `std_math = ["num_traits/std"]` — except maybe supporting it in the future.

vks commented:
I added an "std_math" feature and made sure our CI covers the different `rand_distr` features.

vks commented:
@dhardy I think I addressed all of your concerns?

dhardy reviewed (approved):
Yes, sorry. Approved.
